File: 442_Implement_Trie_(Prefix Tree).py
#!/usr/bin/python3

'''
442. Implement Trie (Prefix Tree)
Implement a trie with insert, search, and startsWith methods.

Example
insert("lintcode")
search("code")
>>> false
startsWith("lint")
>>> true
startsWith("linterror")
>>> false
insert("linterror")
search("lintcode)
>>> true
startsWith("linterror")
>>> true

Notice
You may assume that all inputs are consist of lowercase letters a-z.

Reference
https://zxi.mytechroad.com/blog/data-structure/leetcode-208-implement-trie-prefix-tree/


Author: Yuan Li
Date: 9/1/2018 
Difficulty: Medium
'''

# time: O(l)
# space: O(prefixes) = O(n * l^2)
class Trie:
    
    def __init__(self):
        # do intialization if necessary
        self.root = {}

    """
    @param: word: a word
    @return: nothing
    """
    def insert(self, word):
        # write your code here
        p = self.root
        for c in word:
            if c not in p:
                p[c] = {}
            
            p = p[c]
            
        p['#'] = True

    """
    @param: word: A string
    @return: if the word is in the trie.
    """
    def search(self, word):
        # write your code here
        node = self.find(word)
        return node is not None and '#' in node 

    """
    @param: prefix: A string
    @return: if there is any word in the trie that starts with the given prefix.
    """
    def startsWith(self, prefix):
        # write your code here
        return self.find(prefix) is not None
        
    def find(self, prefix):
        p = self.root
        for c in prefix:
            if c not in p:
                return None
                
            p = p[c]
            
        return p
        

# Date: 12/30/2018      
class TrieNode:
    
    def __init__(self):
        self.children = {}
        self.is_word = False
        
class Trie:
    
    def __init__(self):
        # do intialization if necessary
        self.root = TrieNode()

    """
    @param: word: a word
    @return: nothing
    """
    def insert(self, word):
        # write your code here
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            
            node = node.children[c]
            
        node.is_word = True

    """
    @param: word: A string
    @return: if the word is in the trie.
    """
    def search(self, word):
        # write your code here
        node = self.find(word)
        return node is not None and node.is_word 

    """
    @param: prefix: A string
    @return: if there is any word in the trie that starts with the given prefix.
    """
    def startsWith(self, prefix):
        # write your code here
        return self.find(prefix) is not None
        
    def find(self, prefix):
        node = self.root
        for c in prefix:
            if c not in node.children:
                return None
                
            node = node.children[c]
            
        return node

File: 442_Implement_Trie_(Prefix Tree).cpp
/*
442. Implement Trie (Prefix Tree)
Implement a trie with insert, search, and startsWith methods.

Example
insert("lintcode")
search("code")
>>> false
startsWith("lint")
>>> true
startsWith("linterror")
>>> false
insert("linterror")
search("lintcode)
>>> true
startsWith("linterror")
>>> true

Notice
You may assume that all inputs are consist of lowercase letters a-z.

Reference
https://zxi.mytechroad.com/blog/data-structure/leetcode-208-implement-trie-prefix-tree/


Author: Yuan Li
Date: 9/2/2018 
Difficulty: Medium
*/

// time: O(l)
// space: O(prefixes) = O(n * l^2)
class Trie {
  public:
    Trie() {
        // do intialization if necessary
        root_.reset(new TrieNode()); // smart pointer - no need to delete
    }

    /*
     * @param word: a word
     * @return: nothing
     */
    void insert(const string &word) {
        // write your code here
        TrieNode* p = root_.get();
        for (char c : word) {
            if (!p->children.count(c)) {
                p->children[c] = new TrieNode();
            }
            
            p = p->children[c];
        }
        
        p->word_end = true;
    }

    /*
     * @param word: A string
     * @return: if the word is in the trie.
     */
    bool search(const string &word) const {
        // write your code here
        const TrieNode* node = find(word);
        return node && node->word_end;
    }

    /*
     * @param prefix: A string
     * @return: if there is any word in the trie that starts with the given prefix.
     */
    bool startsWith(const string &prefix) const {
        // write your code here
        return find(prefix) != nullptr;
    }
    
  private:
    struct TrieNode {
        TrieNode() : word_end(false) {} // constructor
        
        ~TrieNode() { // destructor
            for (auto &item : children) {
                if (item.second) {
                    delete item.second;
                }
            }    
        }
        
        bool word_end;
        unordered_map<char, TrieNode*> children;
    };
    
    unique_ptr<TrieNode> root_; // dummy head
    
  private:
    const TrieNode* find(const string &prefix) const {
        TrieNode* p = root_.get();
        for (char c : prefix) {
            if (!p->children.count(c)) {
                return nullptr;
            } 
            
            p = p->children[c];
        }
        
        return p;
    }
};
